serve/mlc_serve/engine/engine_common.py

```diff
@@ -206,10 +169,54 @@ def check_stopping_sequences(stopping_criteria, output_text, delta, is_ended):
     return output_text, delta, is_ended
 
 
-def update_sequence(
+def prepare_logprob(
+    logprob_info,
+    delta,
+    gen_seq: GenerationSequence,
+    prompt_token_ids,
+    tokenizer,
+):
+    if logprob_info is None:
+        return []
+
+    outputs = []
+    for info in logprob_info:
+        assert info is not None
+        assert info.top_token_ids is not None
+        assert info.top_logprobs is not None
+
+        top_logprobs: List[TopLogprobs] = []
+        token_ids = info.top_token_ids.cpu().numpy()
+        logprobs = info.top_logprobs.cpu().numpy()
+
+        for top_token_id, top_logprob in zip(token_ids, logprobs):
+            top_logprobs.append(
+                TopLogprobs(
+                    token=detokenize_incrementally(
```

Comment:
It is not clear for me that it is that client of new OpenAI API expects as output token. @zxybazh could you comment it again with the reference on OpenAI description?

Reply:
Thanks for notifying me of this change! From logprobs example given in OpenAI's [API reference](https://platform.openai.com/docs/api-reference/chat#chat-create-logprobs), the logprobs option is defined as `Whether to return log probabilities of the output tokens or not. If true, returns the log probabilities of each output token returned in the content of message. ` which doesn't explicitly say what their `output tokens` are.

On the other hand, in their example, we can observe some tokens are with prefix of a space like ` there` and ` today`.  Also the output token could be something like `?` or `?\n`. Therefore, I think we can use the the result of decoding based on the context (to preserve the prefix of a single space) like `detokenize_incrementally` function to replace some `tokenizer.decode` function so that we can easily distinguish token strings like ` Here` and `Here`.

Reply:
So I guess we are good here? As discussed with @vvchernov offline, we can add the `token_id` field as a follow-up.